lib/src/back_end/solution_cache.dart

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import '../piece/piece.dart';
+import 'solution.dart';
+import 'solver.dart';
+
+/// Maintains a cache of [Piece] subtrees that have been previously solved.
+///
+/// If a given [Piece] has newlines before and after it, then (in most cases,
+/// assuming there are no other constraints), then the way it is formatted
+/// really only depends on its leading indentation and state. In that case, we
+/// can format that piece using a separate Solver and insert the results in any
+/// Solution that has that piece at that leading indentation.
+///
+/// This cache stores those previously formatted subtree pieces so that
+/// [CodeWriter] can reuse them across Solution[s].
```

Comment:
Nit: The brackets for `Solution[s]` are probably not supposed to be like that.

Reply:
Heh, oops. :)